src/include/index/ivf_pq_index.h

```diff
@@ -385,6 +405,56 @@ class ivf_pq_index {
             std::nullopt,
             num_clusters_,
             temporal_policy_);
+
+    if (index_load_strategy_ ==
```

Comment:
This is a bit confusing. Shouldn't this also have `index_load_strategy_ ==IndexLoadStrategy::DEFAULT ||`

Reply:
Thanks, so there are four cases:
* IndexLoadStrategy::PQ_INDEX, upper_bound = 0
  * Load partitioned_pq_vectors_
* IndexLoadStrategy::PQ_INDEX, upper_bound > 0
  * Do not load partitioned_pq_vectors_
* IndexLoadStrategy::PQ_INDEX_AND_RERANKING_VECTORS, upper_bound = 0
  * Load partitioned_pq_vectors_
* IndexLoadStrategy::PQ_INDEX_AND_RERANKING_VECTORS, upper_bound > 0
  * Throw an error because this is invalid

Before I had `if (index_load_strategy_ == IndexLoadStrategy::PQ_INDEX_AND_RERANKING_VECTORS || upper_bound == 0) {` which still worked, but we can actually simplify to `if (upper_bound == 0) {` and it will do the same thing. So updated, thanks.

Reply:
To make this code/enum logic clearer, I would suggest to load data only based on `IndexLoadStrategy`. `upper_bound>0` should only be relevant/allowed for `IndexLoadStrategy::METADATA_ONLY`.

Reply:
I can if you really want, but I think that leads to a quite verbose API that won't be easy to use (if people start using the C++ API)...

Reply:
I am not sure why it is more verbose for C++ users. We already added the `IndexLoadStrategy` argument. And what I am saying is to reduce complexity by only allowing `upper_bound` only for `IndexLoadStrategy::METADATA_ONLY`. 

Atm  the different enums in that strategy are not well defined. For example, as you have it now, if a user sets `IndexLoadStrategy::METADATA_ONLY` the index is not really usable (can't perform queries), which is not a useful mode.

Reply:
> if a user sets IndexLoadStrategy::METADATA_ONLY the index is not really usable (can't perform queries), which is not a useful mode.

But that's the whole point - it's so that you open an index that will help you query for metadata and nothing else.

Reply:
But sure, I'll update

Reply:
Oh wait, doing that would actually make things quite complicated, as then this check won't work, we'll return early instead of constructing `flat_ivf_centroids_` and `cluster_centroids_`:
```
if (index_load_strategy_ == IndexLoadStrategy::METADATA_ONLY) {
      return;
    }

    flat_ivf_centroids_ =
        tdbPreLoadMatrix<flat_vector_feature_type, stdx::layout_left>(
            group_->cached_ctx(),
            group_->flat_ivf_centroids_uri(),
            std::nullopt,
            num_partitions_,
            0,
            temporal_policy_);

    cluster_centroids_ =
        tdbPreLoadMatrix<flat_vector_feature_type, stdx::layout_left>(
            group_->cached_ctx(),
            group_->cluster_centroids_uri(),
            std::nullopt,
            std::nullopt,
            num_clusters_,
            temporal_policy_);
```
I guess maybe the real issue here is around the definition of `METADATA_ONLY`...

Reply:
Yes the definition of `METADATA_ONLY` is the main problem here, maybe we should call it `OOC`. Opening an index without query capabilities is not really valid IMO.

Reply:
Well the point was to support Python's `open_for_remote_query_execution`:
```
enum class IndexLoadStrategy {
  // Used by Python to support `open_for_remote_query_execution`. In this case
  // we don't want to load any data, we just want to load metadata so that we
  // can call
  // index.dimensions() and other metadata getters.
  METADATA_ONLY,
  // Load the PQ index data as well as metadata.
  PQ_INDEX,
  // Load everything DEFAULT does, but also preload feature vectors for
  // reranking.
  PQ_INDEX_AND_RERANKING_VECTORS,
};
```

Reply:
But we could change to this if you want:
```
enum class IndexLoadStrategy {
  // Will load the index as an OOC index.
  PQ_OOC,
  // Load the PQ index data as well as metadata.
  PQ_INDEX,
  // Load everything DEFAULT does, but also preload feature vectors for
  // reranking.
  PQ_INDEX_AND_RERANKING_VECTORS,
};
```

Reply:
Okay, updated PR.

> I am not sure why it is more verbose for C++ users.
Well, you can see in the commit that to do an OOC query now requires more code than it did before: https://github.com/TileDB-Inc/TileDB-Vector-Search/pull/518/commits/bf8d4c00b1385fceb5dd96b80695b95d99d00dc3#diff-15e9e750c136c14df58a0721c199d0596c1513e422cd3aa33c0f40a5aeec2223L235

Reply:
IMO the `index_load_strategy` argument should be placed before `upper_bound` and default to `PQ_OOC`

Reply:
Sure, updated to move `index_load_strategy` before `upper_bound`.